### PR TITLE
GridMask implementation with BaseImageAugmentationLayer

### LINKED ISSUE #190: Migrate GridMask to use BaseImageAugmentationLayer
https://github.com/keras-team/keras-cv/blob/master/keras_cv/layers/preprocessing/grid_mask.py

Example in RandomShear.

### PR DESCRIPTION
**Pull request by chjort**
GridMask layer has been updated to use BaseImageAugmentationLayer, which closes #190.
As of now, auto-vectorization of GridMask is not functional using BaseImageAugmentationLayer.

Memory and speed optimizations to `fill_utils.py` are also included in this PR to improve GridMask performance.

### CODE REVIEW

**Comment by bhack** [comment]
Have you checked how the memory is going to increase with the input resolution other then the batch size? As I suppose in this version we need to store all the (dense) maskes in memory with the same input resolution for the whole batch.

**Comment by bhack** [comment]
I've not still seen this in details but I would like to collect some opinion about how this have impacted the code readbilty.

**Comment by chjort** [comment]
> Have you checked how the memory is going to increase with the input resolution other then the batch size? As I suppose in this version we need to store all the (dense) maskes in memory with the same input resolution for the whole batch.

Memory also scales with input resolution, but memory consumption does not come from storing the masks as is. It comes from the `logical_and` operation in [this line](https://github.com/keras-team/keras-cv/pull/159/files#diff-089b1d1db30c741129c082ecc77c1fe1860a81788a0f62afe8680a0be6b0355bR52) which computes the masks.

I do not have any numbers for increasing image resolution. Let me know if you would like me to conduct some experiments and collect some numbers. But memory will scale with `O(batch_size * input_resolution)`

**Comment by bhack** [comment]
Have you tried to `jit_compile` It?

**Comment by chjort** [comment]
> Have you tried to `jit_compile` It?

Running the vectorized implementation with `jit_compile=True` on `fill_utils.corners_to_mask` gives minor speedup, but memory consumption remains the same as above:
| Batch sizes 	| Vectorized jit time [ms] 	|
|-------------	|----------------------	|
| 64          	| 1725                 	|
| 128         	| 1981                 	|
| 256         	| 2375                 	|
| 512         	| 3096                 	|

**Comment by bhack** [comment]
Probably the speed-up is a little bit more as we need to subtract the compile time. 
I suppose that also in this version we cannot `jit_compile` the caller for the same reason as in https://github.com/keras-team/keras-cv/pull/146#issuecomment-1049352918 right?

**Comment by LukeWood** [comment]
We are going to lean on the results from #160 to vectorize our KPLs.    It looks like `vectorized_map` actually performs REALLY well in a tf.function, so we will be leaning on this.  Let's see if we can rewrite this layer using the BaseImageAugmentationLayer.

I will be contributing an early sample for how to do this in the near future.

**Comment by LukeWood** [comment]
@chjort the examples are ready.  Check out `RandomShear`, `AutoContrast`, or `RandomSharpness`.

**Comment by chjort** [comment]
Thanks @LukeWood , I will begin looking into this.

**Review by qlzh727 — CHANGES_REQUESTED** [review]
Thanks for the PR.

File: keras_cv/layers/preprocessing/grid_mask.py [hunk 1]
```diff
@@ -75,7 +86,7 @@ def __init__(
         ratio="random",
         rotation_factor=0.15,
         fill_mode="constant",
-        fill_value=0,
+        fill_value=0.0,
         seed=None,
```
[the remark this answers is not in the document]
Comment (chjort) [reply keras_cv/layers/preprocessing/grid_mask.py]: I need `seed` for the `RandomRotation` layer, however `seed` is not a property on `BaseRandomLayer`, so I was forced to access it from the private property `self._random_generator._seed`.

It might be better UX to have it accessible as`self.seed` through parent class, or have random generator be public like `self.random_generator`.

File: keras_cv/layers/preprocessing/grid_mask.py [hunk 2]
```diff
@@ -116,127 +130,102 @@ def _check_parameter_values(self):
...
+        mask = tf.cast(mask, tf.bool)
 
-        if training is None:
-            training = backend.learning_phase()
+        # fill
+        if self.fill_mode == "constant":
+            fill_value = tf.fill(input_shape, self.fill_value)
+        else:
+            # gaussian noise
+            fill_value = tf.random.normal(input_shape)
```
[the remark this answers is not in the document]
Comment (chjort) [reply keras_cv/layers/preprocessing/grid_mask.py]: I agree. It has been moved to `get_random_transformation` now.

**Review by qlzh727 — APPROVED** [review]

**Comment by chjort** [comment]
@qlzh727 Let me know if there are any other changes you would like me to do before this can be merged.

**Comment by chjort** [comment]
> @chjort great work. Is it possible to showcase some demo output?

Absolutely. Here are samples from `GridMask(ratio="random", rotation_factor=0.5, fill_mode="gaussian_noise")`
gridmask

**Comment by LukeWood** [comment]
Thanks @chjort
